### ISSUE TITLE
Can't get Tensorflow 2.0 to properly distribute training across multiple GPUs

### ISSUE THREAD

**Issue by rileypsmith** [tensorflow/tensorflow#46463]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Redhat Linux 7.5
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.0
- Python version: 3.7.7
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: Cuda 10.1/cuDNN 7.6.5
- GPU model and memory: 2x Titan RTX, 24gb each

**Describe the problem**
I have two GPUs, both of which are visible to Tensorflow (and Tensorflow confirms this). However, when I try to train a model using a mirrored strategy it only trains a tiny bit on the second GPU. If I make the first GPU invisible with `export CUDA_VISIBLE_DEVICES=1`, then it will actually train on the second GPU. So I know it is capable of training on both. Furthermore, if I start a model training on the first GPU and then do `export CUDA_VISIBLE_DEVICES=1` and try to train another model, it will train on the second GPU. So I know that Tensorflow can find both GPUs and can train on both GPUs and can have both GPUs training at the same time, but I cannot get it to properly train a single model on both GPUs. 

Running `export CUDA_VISIBLE_DEVICES=0,1` does nothing; Tensorflow will still only train a tiny bit on the second GPU. 

**Provide the exact sequence of commands / steps that you executed before running into the problem**
As stated above, I first ran `export CUDA_VISIBLE_DEVICES=0,1`. I have also tried training without running this command first. The exact code I am using to build my model is below. I have a convolutional neural network called FullModel():
```python
strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    net = FullModel()

for epoch in range(100):
    # Custom training loop
```
I have also tried the following:
```python
strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])
with strategy.scope():
    net = FullModel()
```

**Any other info / logs**
The output of `nvidia-smi` right after beginning training:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.82       Driver Version: 440.82       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN RTX           Off  | 00000000:1B:00.0 Off |                  N/A |
| 52%   70C    P2   199W / 280W |  23537MiB / 24220MiB |     51%      Default |
+-------------------------------+----------------------+----------------------+
|   1  TITAN RTX           Off  | 00000000:68:00.0 Off |                  N/A |
| 41%   47C    P8     4W / 280W |    262MiB / 24219MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     14382      C   python                                     23525MiB |
|    1      3337      G   /usr/bin/X                                    27MiB |
|    1      4783      G   /usr/bin/gnome-shell                          58MiB |
|    1     14382      C   python                                       163MiB |
+-----------------------------------------------------------------------------+
```
Notice that job 14382 IS doing something on GPU 1. But it is only taking up 163mb on that GPU. I thought that maybe I just could train in larger batches or something to maximize my available GPU memory but that doesn't seem to change the situation. For some reason it just isn't using all of that GPU.

The output of `nvidia-smi` after I run `export CUDA_VISIBLE_DEVICES=1` and start another job training:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.82       Driver Version: 440.82       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN RTX           Off  | 00000000:1B:00.0 Off |                  N/A |
| 53%   71C    P2   136W / 280W |  23537MiB / 24220MiB |     43%      Default |
+-------------------------------+----------------------+----------------------+
|   1  TITAN RTX           Off  | 00000000:68:00.0 Off |                  N/A |
| 41%   62C    P2   246W / 280W |  23499MiB / 24219MiB |     75%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     14382      C   python                                     23525MiB |
|    1      3337      G   /usr/bin/X                                    27MiB |
|    1      4783      G   /usr/bin/gnome-shell                          58MiB |
|    1     14382      C   python                                       163MiB |
|    1     19376      C   python                                     23237MiB |
+-----------------------------------------------------------------------------+
```

**Comment by amahendrakar** [comment]
@rileypsmith,
Could you please update TensorFlow to the latest stable version v2.4 and check if you are facing the same issue? Thanks!

**Comment by rileypsmith** [comment]
@amahendrakar 
I do not have permission on this machine to update cuda to the proper installation which is why I have refrained from updating my tensorflow. Are there any ways to solve this problem for tensorflow 2.0? I will try to get the proper version of cuda on the machine to update my tensorflow but I may not be able to

**Comment by amahendrakar** [comment]
@rileypsmith,
CUDA 10.1 and cuDNN 7.6 is compatible with TensorFlow v2.3 as well. Is it possible for you to update the TensorFlow package?

Also in order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and the dataset you are using. Thanks!

**Comment by rileypsmith** [comment]
Apologies for the delayed response. I did not notice that Tensorflow v2.3 is also compatible with CUDA 10.1 and cuDNN 7.6. That is my mistake. I will try updating Tensorflow.

My complete code to reproduce this issue is incredibly lengthy and I cannot post all of it here. Below is my model and training loop (my real model is more complex and spans several files but is similar in principle to that which is shown below and the issue is robust to changes in model anyways).

Model:
```python
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import Conv2D, UpSampling2D, ReLU

class FullModel(Model):

    def __init__(self, use_bias=True, **kwargs):
        super(AutoEncoder, self).__init__(**kwargs)

        self.conv1 = Conv2D(128, kernel_size=(3, 3), strides=(2, 2), padding='same', use_bias=use_bias)
        self.conv2 = Conv2D(128, kernel_size=(3, 3), strides=(2, 2), padding='same', use_bias=use_bias)
        self.conv3 = Conv2D(128, kernel_size=(3, 3), strides=(2, 2), padding='same', use_bias=use_bias)
        self.conv4 = Conv2D(128, kernel_size=(3, 3), strides=1, padding='same', use_bias=use_bias)
        self.conv5 = Conv2D(128, kernel_size=(3, 3), strides=1, padding='same', use_bias=use_bias)
        self.conv6 = Conv2D(3, kernel_size=(3, 3), strides=1, padding='same', use_bias=use_bias)

        self.upsample = UpSampling2D()
        self.relu = ReLU()

    def call(self, x):
        x = self.conv1(x)
        x = self.relu(x)

        x = self.conv2(x)
        x = self.relu(x)

        x = self.conv3(x)
        x = self.relu(x)

        x = self.upsample(x)
        x = self.conv4(x)
        x = self.relu(x)

        x = self.upsample(x)
        x = self.conv5(x)
        x = self.relu(x)

        x = self.upsample(x)
        x = self.conv6(x)

        return x
```

Training:
```python
import tensorflow as tf
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import Progbar

from models import FullModel

def train_step(net, opt, batch):
    # Split the batch into current and previous frames
    curr, prev = split_batch(batch)

    # Start tracking gradient
    with tf.GradientTape() as tape:
        # Put batch through network
        pred = net(curr, prev)
        # Compute loss as MSE
        loss = tf.math.reduce_mean(tf.math.square(pred - curr))
        # Setup gradient
        gradient = tape.gradient(loss, net.trainable_variables)
        # Apply gradients
        opt.apply_gradients(zip(gradient, net.trainable_variables))

    # Return loss so it can be tracked during training
    return pred, loss

def train(epochs=100, steps_per_epoch_train=0, train_data=None, batch_size=16):
    """
    Function to run through the dataset for the desired number of epochs
    """
    assert train_data

    # Build model
    strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])
    with strategy.scope():
        net = FullModel()

    # Build optimizer
    opt = Adam(learning_rate=3e-5)

    for epoch in range(epochs):
        print(f'Epoch {epoch+1}/{epochs}')

        # Setup progress bar
        pb = Progbar(steps_per_epoch_train)

        # Set summary statistics
        train_loss = 0
        num_batches = 0

        # Loop through training set
        for i, batch in train_data.enumerate():
            if i > steps_per_epoch_train:
                break
            # Do one step of SGD for each batch
            loss = train_step(net, opt, batch)
            # Update summary statistics
            train_loss += loss
            num_batches += 1
            # Setup train loss for monitoring during training
            values = [
                ('Train Loss', train_loss / num_batches)
            ]
            pb.add(1, values=values)
```
Very simple. My dataset is just consecutive frames from a video. I split a video up into frames, reshape them, and then pass them as a batch. So each batch is a tensor of shape `(batch_size, 2, 256, 256, 3)`. `batch_size` is the number of different videos, size 2 along axis 1 is because I have the previous and current frame for each video, and then `256, 256, 3` is just the reshaped rgb image of each frame.

However, this issue is robust to changes in dataset, model, and training. I have been able to observe the same issue when using the TensorFlow functional API to build a simple model and train with `.fit()`  on a regular dataset containing examples and labels.

**Comment by rileypsmith** [comment]
I updated Tensorflow to version 2.3 as suggested and this solved my issue! Thank you for your help. I do believe the issue I was experiencing in version 2.0 was a bug of sorts because there were no issues in my setup but I am closing this issue since I am now able to train on multiple GPUs in tversion 2.3.

Also, just for reference because this tripped me up for a second: I am running out of a conda virtual environment but I had to install tensorflow using `pip install tensorflow=2.3`. The current conda installation of tensorflow-gpu is version 2.2 for Linux.
